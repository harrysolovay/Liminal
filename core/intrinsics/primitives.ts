import { assert } from "@std/assert"
import type { Type } from "../Type.ts"
import { declare } from "./declare.ts"

export { null_ as null }
const null_: Type<null> = declare({
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean> = declare({
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number> = declare({
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number> = declare({
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string> = declare({
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
