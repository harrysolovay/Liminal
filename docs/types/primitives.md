# Primitive Types

```ts twoslash
import { L } from "liminal"
// ---cut---
L.boolean
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.number
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.integer
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.string
// ^?
```

<br />
