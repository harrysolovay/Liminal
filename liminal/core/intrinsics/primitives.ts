import { assert } from "@std/assert"
import { declare } from "../declare.ts"
import type { Type } from "../Type.ts"

export { null_ as null }
const null_: Type<null, never> = declare({
  jsonType: "null",
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean, never> = declare({
  jsonType: "boolean",
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number, never> = declare({
  jsonType: "integer",
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number, never> = declare({
  jsonType: "number",
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string, never> = declare({
  jsonType: "string",
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
