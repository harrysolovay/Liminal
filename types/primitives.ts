import { Ty } from "./Ty.ts"

export const none: Ty<null, never> = Ty(() => ({
  type: "null",
}))

export const boolean: Ty<boolean, never> = Ty(() => ({
  type: "boolean",
}))

export const number: Ty<number, never> = Ty(() => ({
  type: "number",
}))

export const string: Ty<string, never> = Ty(() => ({
  type: "string",
}))
