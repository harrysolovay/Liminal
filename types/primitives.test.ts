import { assertTySnapshot } from "test_util"
import { boolean, none, number, string } from "./primitives.ts"

Deno.test("boolean", async (t) => {
  await assertTySnapshot(t, boolean)
})

Deno.test("none", async (t) => {
  await assertTySnapshot(t, none)
})

Deno.test("number", async (t) => {
  await assertTySnapshot(t, number)
})

Deno.test("string", async (t) => {
  await assertTySnapshot(t, string)
})
