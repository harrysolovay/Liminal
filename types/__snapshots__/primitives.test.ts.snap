export const snapshot = {};

snapshot[`boolean 1`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      type: "boolean",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`boolean 2`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Description one.",
      type: "boolean",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`boolean 3`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Description two. Description one.",
      type: "boolean",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`none 1`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      type: "null",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`none 2`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Description one.",
      type: "null",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`none 3`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Description two. Description one.",
      type: "null",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`number 1`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      type: "number",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`number 2`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Description one.",
      type: "number",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`number 3`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Description two. Description one.",
      type: "number",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`string 1`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      type: "string",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`string 2`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Description one.",
      type: "string",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`string 3`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Description two. Description one.",
      type: "string",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;
